discoverd/client/client.go

```diff
@@ -1,266 +1,205 @@
 package discoverd
 
 import (
-	"encoding/json"
 	"errors"
 	"fmt"
+	"net"
 	"net/http"
 	"os"
 	"strings"
+	"sync"
 	"time"
 
 	dt "github.com/flynn/flynn/discoverd/types"
 	"github.com/flynn/flynn/pkg/dialer"
 	"github.com/flynn/flynn/pkg/httpclient"
-	hh "github.com/flynn/flynn/pkg/httphelper"
 	"github.com/flynn/flynn/pkg/stream"
 )
 
-type Service interface {
-	Leader() (*Instance, error)
-	Instances() ([]*Instance, error)
-	Addrs() ([]string, error)
-	Leaders(chan *Instance) (stream.Stream, error)
-	Watch(events chan *Event) (stream.Stream, error)
-	GetMeta() (*ServiceMeta, error)
-	SetMeta(*ServiceMeta) error
-	SetLeader(string) error
-}
-
 var ErrTimedOut = errors.New("discoverd: timed out waiting for instances")
+var PinTTL = 60 * time.Second
+
+type Config struct {
+	Endpoints []string
+}
 
 type Client struct {
-	c *httpclient.Client
+	servers    []*httpclient.Client
+	pinned     int
+	pinUpdated time.Time
+	leader     int
+	mu         sync.RWMutex
 }
 
-func NewClient() *Client {
-	url := os.Getenv("DISCOVERD")
-	if url == "" {
-		url = "http://127.0.0.1:1111"
+func NewClientWithConfig(config Config) *Client {
+	client := &Client{
+		servers: make([]*httpclient.Client, 0, len(config.Endpoints)),
+	}
+	for _, e := range config.Endpoints {
+		client.servers = append(client.servers, client.httpClient(e))
 	}
-	return NewClientWithURL(url)
+	return client
 }
 
 func NewClientWithURL(url string) *Client {
-	if !strings.HasPrefix(url, "http") {
-		url = "http://" + url
-	}
-	return &Client{
-		c: &httpclient.Client{
-			URL: url,
-			HTTP: &http.Client{
-				Transport:     &http.Transport{Dial: dialer.Retry.Dial},
-				CheckRedirect: redirectPreserveHeaders,
-			},
-		},
-	}
+	return NewClientWithConfig(Config{Endpoints: formatURLs([]string{url})})
```

Comment:
`formatURLs` should take `urls ...string` so you can call it here like `formatURLs(url)`